### ISSUE TITLE
no support for RedisCluster setups 

### ISSUE THREAD

**Issue by landryb** [nextcloud/notify_push#65]
i have nextcloud instances which dont directly use the plain redis support, but via RedisCluster:
```
  'redis.cluster' => 
  array (
    'seeds' => 
    array (
      0 => 'db1:6380',
      1 => 'db1:6381',
      2 => 'db1:6382',
      3 => 'db2:6380',
      4 => 'db2:6381',
      5 => 'db2:6382',
    ),
    'password' => 'xxx',
    'timeout' => 0.0,
    'read_timeout' => 0.0,
    'failover_mode' => 1,
  ),
```

trying to enable `notify_push` on such setup will miserably blow:
```
$occ app:enable notify_push
[sudo] password for landry: 
An unhandled exception has been thrown:
TypeError: Argument 1 passed to OCA\NotifyPush\Queue\RedisQueue::__construct() must be an instance of Redis, instance of RedisCluster given, called in /data/web/www/nextcloud/apps/notify_push/lib/AppInfo/Application.php on line 62 and defined in /data/web/www/nextcloud/apps/notify_push/lib/Queue/RedisQueue.php:29
Stack trace:
#0 /data/web/www/nextcloud/apps/notify_push/lib/AppInfo/Application.php(62): OCA\NotifyPush\Queue\RedisQueue->__construct(Object(RedisCluster))
```

**Comment by landryb** [comment]
what is strange is that i dunno if the redis used by `notify_push` should be the same as the one used by nextcloud (ie used for message passing ?) or if as a workaround i can point notify_push at another single-instance redis process..

**Comment by mamatt** [comment]
I can confirm that the install failed with a RedisCluster setup

 ```
sudo -u www-data ./occ app:install notify_push
An unhandled exception has been thrown:
TypeError: Argument 1 passed to OCA\NotifyPush\Queue\RedisQueue::__construct() must be an instance of Redis, instance of RedisCluster given, called in /var/www/nextcloud/apps/notify_push/lib/AppInfo/Application.php on line 62 and defined in /var/www/nextcloud/apps/notify_push/lib/Queue/RedisQueue.php:29
```
```
'redis.cluster' => 
  array (
    'seeds' => 
    array (
      0 => '172.26.0.165:7000',
      1 => '172.26.0.165:7001',
      2 => '172.26.0.166:7000',
      3 => '172.26.0.166:7001',
      4 => '172.26.0.167:7000',
      5 => '172.26.0.167:7001',
      6 => '172.26.0.168:7000',
      7 => '172.26.0.168:7001',
    ),
    'timeout' => 0.0,
    'read_timeout' => 0.0,
    'failover_mode' => 1,
  ),
```

**Comment by l00v3** [comment]
Hi! If that is true, we will have problem implementing Files Notify Push backend, if Redis Cluster is not working. Hope the support for Redis Cluster will be added soon : )

**Comment by wiswedel** [comment]
related: https://github.com/nextcloud/notify_push/issues/44

**Comment by icewind1991** [comment]
redis cluster support should work now in master and I would appreciate some testing before I tag a release

You can find a build of the app in the [github actions](https://github.com/nextcloud/notify_push/actions/runs/702066737)

**Comment by l00v3** [comment]
Tested on latest NC v21.0.1 with above notify_push build and from initial tests it works as it should : ) thank you very much! Will test it further and provide feedback if I notice something.
Greetings

**Comment by icewind1991** [comment]
released with 0.1.7

**Comment by landryb** [comment]
i've tested this against a rediscluster setup, it looks improved but chokes on this config:
```
  'redis.cluster' => [
     'seeds' => [
       'db1:6380',
       'db1:6381',
       'db1:6382',
       'db2:6380',
       'db2:6381',
       'db2:6382',
     ],
     'password' => 'xxx',
     'timeout' => 0.0,
     'read_timeout' => 0.0,
     'failover_mode' => \RedisCluster::FAILOVER_ERROR
  ],
```
the service complains at startup:
```
Apr 22 16:27:31 web1 notify_push[33086]:    0: Failed to parse config
Apr 22 16:27:31 web1 notify_push[33086]:    1: Failed to parse config file
Apr 22 16:27:31 web1 notify_push[33086]:    2:
Apr 22 16:27:31 web1 notify_push[33086]:       .. |
Apr 22 16:27:31 web1 notify_push[33086]:       41 |      'timeout' => 0.0,
Apr 22 16:27:31 web1 notify_push[33086]:       42 |      'read_timeout' => 0.0,
Apr 22 16:27:31 web1 notify_push[33086]:       43 |      'failover_mode' => \RedisCluster::FAILOVER_ERROR
Apr 22 16:27:31 web1 notify_push[33086]:          |                         ^^ No valid token found, expected one of [Bool, Integer, Float, LiteralString, Null, Array, SquareOpen]
Apr 22 16:27:31 web1 notify_push[33086]:       44 |   ],
```

ofc, i can replace `\RedisCluster::FAILOVER_ERROR` by its numeric equivalent (eg 1) but well .. that should be a valid config :)

with that, the systemd service starts fine, and i can run `occ notify_push:setup` fine on this clustered redis instance.

but alas, issue #62 is still valid, creating a file via webdav is properly detected by one of the systemd services:
```
[2021-04-22 16:38:55.396119 +02:00] DEBUG [notify_push] src/lib.rs:399: Received storage update notification for storage 48 and path files/logo.svg
```
but zero notification is sent to a client browsing `/apps/files/`, he wont get a notification that `logo.svg` was created, nor will see it until he refreshes the web page. there's still no websocket connection seen in the web console.

anyway, thanks for adding RedisCluster support :)
